Seperate js out template to avoid render issue

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Main purpose of this PR to seperate js out of template avoid mix js-in-php

Previously if we set store to production with minify enabled. When access backend will raise error because phtml can't render properly

Browser screen return 
"An error has happened during application run. See exception log for details."

Reference var/log/system.log

Here is the error log when we meet steps scenario
main.CRITICAL: ParseError: syntax error, unexpected '<<' (T_SL) in /var/www/html/var/view_preprocessed/pub/static/app/code/Magento/Backend/view/adminhtml/templates/store/switcher.phtml:44
Stack trace:
0 /var/www/html/lib/internal/Magento/Framework/View/Element/Template.php(273): Magento\Framework\View\TemplateEngine\Php->render()
1 /var/www/html/generated/code/Magento/Backend/Block/Store/Switcher/Interceptor.php(428): Magento\Framework\View\Element\Template->fetchView()

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Switch to production mode
bin/magento deploy:mode:set production
2. Run command to config minify html
bin/magento config:set --lock-env dev/template/minify_html 1
3. Clear cache and run deploy static content for backend area
4. Navigate to admin and login with credentials


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#32454: Seperate js out template to avoid render issue

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/web/js/store-switcher.js:
```diff
@@ -0,0 +1,88 @@
+], function(jQuery, $t, confirm) {
+    'use strict';
+
+    (function($) {
```
I have feeling that it should be inside the `return function(config) { ...`. What do you thing?

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/web/js/store-switcher.js:
```diff
@@ -0,0 +1,88 @@
+define([
+    'jquery',
+    'mage/translate',
+    'Magento_Ui/js/modal/confirm'
+], function(jQuery, $t, confirm) {
```
I believe it's better to use short alias of jQuery in all places. What do you thing?
```suggestion
define([
    'jquery',
    'mage/translate',
    'Magento_Ui/js/modal/confirm'
], function($, $t, confirm) {
```

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/templates/store/switcher.phtml:
```diff
@@ -8,222 +8,132 @@
+        {
+            "#store-switcher-backend": {
+                "Magento_Backend/js/store-switcher": {
+                    "useConfirm": <?= /* @noEscape */ $block->getUseConfirm(); ?>,
```
Seems like it is bool, so it will be not be shown correctly here.

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/templates/store/switcher.phtml:
```diff
@@ -8,222 +8,132 @@
+            "#store-switcher-backend": {
+                "Magento_Backend/js/store-switcher": {
+                    "useConfirm": <?= /* @noEscape */ $block->getUseConfirm(); ?>,
+                    "isUsingIframe": <?= /* @noEscape */ $block->isUsingIframe(); ?>,
```
Seems like it is bool, so it will be not be shown correctly here.

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/templates/store/switcher.phtml:
```diff
@@ -8,222 +8,132 @@
+                    "useConfirm": <?= /* @noEscape */ $block->getUseConfirm(); ?>,
+                    "isUsingIframe": <?= /* @noEscape */ $block->isUsingIframe(); ?>,
+                    "switchUrl": "<?= $block->escapeUrl($block->getSwitchUrl()); ?>",
+                    "storeId": <?= /* @noEscape */ $block->getStoreId(); ?>
```
I bevieve we should convert it to int before printing

mrtuvn replied:
modified as your suggestion

ihor-sviziev commented:
@mrtuvn please add reference to https://github.com/magento/magento2/commit/c8a50131c829529a0bdd1dae8b7cfa8226293b35#commitcomment-48119352

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/layout/adminhtml_dashboard_index.xml:
```diff
@@ -10,9 +10,9 @@
-                <action method="setUseConfirm">
-                    <argument name="params" xsi:type="string">0</argument>
-                </action>
+                <arguments>
+                    <argument name="is_using_iframe" xsi:type="string">0</argument>
+                </arguments>
```
Please explain why you did it?

mrtuvn replied:
This variable not set anywhere in code-base so i think we should declare it here

mrtuvn replied:
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Backend/Block/Store/Switcher.php#L614
But it called in this class

mrtuvn replied:
Add default value here. Other modules core or 3rd-party providers extensions can easy override it via design area (adminhtml) by layout

mrtuvn replied:
setUseConfirm is unneccesary because it's already set via Block

ihor-sviziev replied:
It feels like it supposes to change the following flag: 
https://github.com/magento/magento2/pull/32434/files#diff-5c07e420bed1b6e2edda1d6165e1ca2dc4eacaab999bf149eaffa4f53c3823afR47

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/web/js/store-switcher.js:
```diff
@@ -0,0 +1,88 @@
+                scopeSwitcherHandler(switcherParams);
+            } else {
+                if (config.useConfirm) {
+                    confirm({
+                        content: $t('Please confirm scope switching. All data that hasn\'t been saved will be lost.'),
+                        actions: {
+                            confirm: function () {
+                                reload();
+                            },
+                            cancel: function () {
+                                obj.value = config.storeId ? config.storeId : '';
+                            }
+                        }
+                    });
```
maybe that's a bit stupid question, but can we wrap it to require and load `'mage/translate'` and `'Magento_Ui/js/modal/confirm'` components when this really needed? It should a bit improve page load time

mrtuvn replied:
updated

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/layout/adminhtml_dashboard_index.xml:
```diff
@@ -10,9 +10,10 @@
-                <action method="setUseConfirm">
-                    <argument name="params" xsi:type="string">0</argument>
-                </action>
+                <arguments>
+                    <argument name="use_confirm" xsi:type="string">1</argument>
+                    <argument name="is_using_iframe" xsi:type="string">0</argument>
+                </arguments>
```
Feels like it won't work, as we don't have this argument in constructor

mrtuvn replied:
you can get it via Block class no problem by getData('variable')

mrtuvn replied:
https://devdocs.magento.com/guides/v2.4/frontend-dev-guide/layouts/xml-instructions.html

mrtuvn commented:
@magento run all tests

ihor-sviziev commented:
Looks good to me. Let's wait for the test results.
Note: I don't think there is a way to cover this use case with automated tests, so I added a label that tests are not required.

mrtuvn commented:
@magento run Database Compare, Functional Tests CE, Functional Tests B2B, Functional Tests EE, Integration Tests, Static Tests, Unit Tests

sivaschenko commented:
@magento create issue

sivaschenko commented:
@mrtuvn can you please add the details on the actual result, i.e. the error that you are getting to the description

ihor-sviziev commented:
@sivaschenko i see now in description we have following:
> Here is the error log when we meet steps scenario
ParseError: syntax error, unexpected '<<' (T_SL) in /var/www/html/var/view_preprocessed/pub/static/app/code/Magento/Backend/view/adminhtml/templates/store/switcher.phtml

from my perspective it should be p2 priority

mrtuvn commented:
> BTW @mrtuvn it’s not clear to me why the phtml file was copied to view_preprocessed folder. It seems wrong to me. Isn’t it?

Because i used setting dev/static/sign = 1 ( So i think that function work as we expected). Should we correct use this setting in case production ?

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Backend/Block/Store/Switcher.php:
```diff
@@ -616,4 +616,17 @@ public function isUsingIframe()
         }
         return false;
     }
+
+    /**
+     * Get whether confirm popup is being used
+     *
+     * @return bool
+     */
+    public function getUseConfirm()
+    {
+        if ($this->hasData('use_confirm')) {
+            return (bool)$this->getData('use_confirm');
+        }
+        return false;
+    }
```
This class is marked as API, so we can't add new methods due to the backward compatibility policy.

ihor-sviziev commented on app/code/Magento/Backend/Block/Store/Switcher.php:
```diff
@@ -113,8 +114,7 @@ public function __construct(
     protected function _construct()
     {
         parent::_construct();
-
-        $this->setUseConfirm(true);
```
maybe we can add a check if not set - then set it to true?

mrtuvn replied:
Maybe i should add check like this 

```php 
if (!$this->hasData('use_confirm')) {
            $this->setUseConfirm(true);
        } elseif ($this->hasData('use_confirm')) {
            $this->setUserConfirm((bool)$this->getData('use_confirm'));
        }
```

and remove public additional method

ihor-sviziev replied:
Maybe like this?
```php
$this->setUserConfirm($this->hasData('use_confirm') ? (bool)$this->getData('use_confirm') : true);
```

mrtuvn replied:
Ok i have updated that

ihor-sviziev commented:
> Because i used setting (Sign Static Files) dev/static/sign = 1 ( So i think that function deploy static worked as we expected). Should we correct use this setting in case production ?

Just double-checked - it works in this way only in case of HTML minification.

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/templates/store/switcher.phtml:
```diff
@@ -8,222 +8,132 @@
-                    }
+    <script type="text/x-magento-init">
+        {
+            ".store-switcher": {
```
Looks like it should be one global service,
 ```suggestion
            "*": {
```

mrtuvn replied:
Corrected that too

ihor-sviziev commented:
Looks good. Let's wait for the test results
@magento run all tests

mrtuvn commented:
@magento run Functional Tests CE

ihor-sviziev commented on app/code/Magento/Backend/view/adminhtml/web/js/store-switcher.js:
```diff
@@ -0,0 +1,124 @@
+
+                scopeSwitcherHandler(switcherParams);
+            } else if (config.useConfirm) {
+                require([
+                    'mage/translate',
+                    'Magento_Ui/js/modal/confirm'
+                ], function ($t, confirm) {
```
I believe moving to a separate require section broke functional tests, as this confirmation popup appears not immediately anymore.
There two options:
1. fix the failing action groups
2. move back these dependencies to the global `define` block of this file.

Actually, I'm fine with both options.

mrtuvn replied:
i have updated follow option 2 for minimal function break and less impact to performance when backend we don't need to serious about performance

ihor-sviziev commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests B2B

mrtuvn commented:
re-run tests
@magento run all tests

mrtuvn commented:
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE, Integration Tests, Static Tests, Unit Tests

mrtuvn commented:
@magento run Functional Tests CE

mrtuvn commented:
@magento run Static Tests, Functional Tests CE, Functional Tests B2B, Functional Tests EE

mrtuvn commented:
@magento run Functional Tests CE

mrtuvn commented on app/code/Magento/Backend/Block/Store/Switcher.php:
```diff
@@ -114,7 +115,8 @@ protected function _construct()
-        $this->setUseConfirm(true);
+        $this->setUserConfirm($this->hasData('use_confirm') ? (bool)$this->getData('use_confirm') : true);
+
         $this->setUseAjax(true);
```
i got silly wrong typo here @@ setUserConfirm => setUseConfirm
LOL

mrtuvn committed:
Seperate js out template to avoid render issue


update


update


update


improve script load


update


update


update


fix static tests


fix static tests


update


update


update


update


update


update


update


update


update


update


update


up


update


fix wrong typo

mrtuvn commented:
Eventually! Let's wait for final results
@magento run all tests

mrtuvn commented:
@magento run Unit Tests, Functional Tests EE

mrtuvn commented:
last fail related external module

MC-31667: Validate Banner with Collage Right Appearance with No Fallback Image, Disabled Infinite Loop, Disabled Play Only When Visible, and Disabled Lazy Load
app/code/Magento/PageBuilder/Test/Mftf/Test/AdminPageBuilderBannerCollageRightAppearanceTest/BannerCollageRightVideoBackgroundNoFallbackImageDisabledLoopAndPlayWhenVisibleAndLazyLoad.xml

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8928](https://jira.corp.magento.com/browse/ENGCOM-8928) has been created to process this Pull Request

mrtuvn commented:
@magento run Functional Tests EE

engcom-Oscar commented:
:heavy_check_mark: QA passed

Was checked case from Manual testing scenarios 
1. Switch to production mode
bin/magento deploy:mode:set production
2. Run command to config minify html
bin/magento config:set --lock-env dev/template/minify_html 1
3. Clear cache and run deploy static content for backend area
4. Navigate to admin and login with credentials

:x: **Before:**
![Screenshot from 2021-04-08 14-45-20](https://user-images.githubusercontent.com/60198413/114021719-a9abb780-9879-11eb-8461-31384ae9f218.png)

:heavy_check_mark: **After:**
Admin logged-in successfully
![Screenshot from 2021-04-08 14-45-50](https://user-images.githubusercontent.com/60198413/114021734-af090200-9879-11eb-8f19-057c0a94cead.png)
